Dmf/Modules.Library/Dmf_String.c

```diff
@@ -216,8 +287,202 @@ String_WideStringToMultiString(
+    ASSERT(NarrowString != NULL);
+
+    WCHAR* wideString;
+
+    // Create a converted string.
+    //
+    wideString = String_MultiStringToWideString(NarrowString);
+    if (NULL == wideString)
+    {
+        ntStatus = STATUS_UNSUCCESSFUL;
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "String_MultiStringToWideString");
+        goto Exit;
+    }
+
+    // Get length of converted string.
+    //
+    size_t stringLength = wcslen(wideString);
+
+    // Check to make sure that destinations ANSI string's buffer is big enough for 
+    // the string and zero terminator.
+    //
+    if (stringLength + sizeof(WCHAR) > UnicodeString->MaximumLength)
+    {
+        ntStatus = STATUS_BUFFER_TOO_SMALL;
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "UnicodeString's buffer is too small");
+        goto Exit;
+    }
+
+    // Copy the converted string to the destination buffer.
+    //
+    wcscpy_s(UnicodeString->Buffer,
+             UnicodeString->MaximumLength,
+             wideString);
+
+    // Zero-terminate the destination string.
+    //
+    UnicodeString->Buffer[stringLength] = L'\0';
+
+    // Update the length of the new Unicode string.
+    //
+    UnicodeString->Length = (USHORT)(stringLength * sizeof(WCHAR));
+
+    ntStatus = STATUS_SUCCESS;
+
+Exit:
+
+    // Free temporary converted string if it could be allocated.
+    //
+    if (wideString != NULL)
+    {
+        free(wideString);
+    }
+
+    // AnsiString has the converted string.
+    //
+    FuncExit(DMF_TRACE, "ntStatus=%!STATUS!", ntStatus);
+
+    return ntStatus;
+}
+
+#pragma code_seg("PAGE")
+static
+NTSTATUS
+String_WideStringCopyAsAnsi(
+    _Out_ ANSI_STRING* AnsiString,
+    _In_z_ WCHAR* WideString
+    )
+/*++
+
+Routine Description:
+
+    Copy a Wide String as an Ansi String.
+
+Arguments:
+
+    AnsiString - Target Ansi String.
+    WideString - Source Wide String.
+
+Return Value:
+
+    NTSTATUS
+
+--*/
+{
+    NTSTATUS ntStatus;
+
+    PAGED_CODE();
+
+    FuncEntry(DMF_TRACE);
+
+    ASSERT(AnsiString != NULL);
+    ASSERT(WideString != NULL);
+
+    char* multibyteString;
+
+    // Create a converted string.
+    //
+    multibyteString = String_WideStringToMultiString(WideString);
+    if (NULL == multibyteString)
+    {
+        ntStatus = STATUS_UNSUCCESSFUL;
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "String_WideStringToMultiString");
+        goto Exit;
+    }
+
+    // Get length of converted string.
+    //
+    size_t stringLength = strlen(multibyteString);
+
+    // Check to make sure that destinations ANSI string's buffer is big enough for 
+    // the string and zero terminator.
+    //
+    if (stringLength + sizeof(CHAR) > AnsiString->MaximumLength)
+    {
+        ntStatus = STATUS_BUFFER_TOO_SMALL;
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "AnsiString's buffer is too small");
+        goto Exit;
+    }
+
+    // Copy the converted string to the destination buffer.
+    //
+    strncpy_s(AnsiString->Buffer,

```

Comment:
I believe this returns an error code we could check.

Reply:
I will look at this next update.